File: server_nodejs/src/api/cart/cart.service.js
const { AppError } = require('../../common/errors/AppError');
const db = require('../../models');
const { v4 } = require('uuid');
module.exports = {
    getCart: async (client) => {
        try {
            const cart = await db.sequelize.query(
                `
                SELECT it.name, quantity, link as primary_img, it.price, ct.itemid
                FROM cart_items ct
                    JOIN items it ON ct.itemid = it.uid
                    LEFT JOIN item_imgs itm ON it.uid = itm.itemid AND itm.is_primary = 1
                WHERE ct.clientid = :clientid AND it.stock_qty >= ct.quantity
                `,
                {
                    replacements: {
                        clientid: client,
                    },
                    type: db.sequelize.QueryTypes.SELECT,
                },
            );
            const itemOutOfStock = await db.sequelize.query(
                `
                SELECT it.name, quantity, link as primary_img, it.price, ct.itemid
                FROM cart_items ct
                    JOIN items it ON ct.itemid = it.uid
                    LEFT JOIN item_imgs itm ON it.uid = itm.itemid AND itm.is_primary = 1
                WHERE ct.clientid = :clientid AND it.stock_qty < ct.quantity
                `,
                {
                    replacements: {
                        clientid: client,
                    },
                    type: db.sequelize.QueryTypes.SELECT,
                },
            );
            const total = cart.reduce((acc, item) => {
                return acc + item.price * item.quantity;
            }, 0);
            return {
                statusCode: 200,
                message: 'Get cart successfully',
                data: {
                    cart,
                    total: parseFloat(total.toFixed(2)),
                    out_of_stock: itemOutOfStock,
                },
            };
        } catch (err) {
            throw new AppError(err.statusCode, err.message);
        }
    },
    updateCart: async (client, item, quantity) => {
        try {
            const cartItem = await db.cart_item.findOne({
                where: {
                    clientid: client,
                    itemid: item,
                },
            });
            if (!cartItem) throw new AppError(404, 'Item not found');
            if (quantity < 1) await cartItem.destroy();
            else {
                const itemObj = await db.item.findOne({
                    where: {
                        uid: item,
                    },
                });
                if (itemObj.stock_qty < quantity) {
                    throw new AppError(400, 'Quantity is greater than stock');
                } else cartItem.quantity = quantity;
            }
            await cartItem.save();
            const cart = await db.sequelize.query(
                `
                SELECT it.name, quantity, link as primary_img, it.price, ct.itemid
                FROM cart_items ct
                    JOIN items it ON ct.itemid = it.uid
                    LEFT JOIN item_imgs itm ON it.uid = itm.itemid AND itm.is_primary = 1
                WHERE ct.clientid = :clientid AND it.stock_qty >= ct.quantity
                `,
                {
                    replacements: {
                        clientid: client,
                    },
                    type: db.sequelize.QueryTypes.SELECT,
                },
            );
            const itemOutOfStock = await db.sequelize.query(
                `
                SELECT it.name, quantity, link as primary_img, it.price, ct.itemid
                FROM cart_items ct
                    JOIN items it ON ct.itemid = it.uid
                    LEFT JOIN item_imgs itm ON it.uid = itm.itemid AND itm.is_primary = 1
                WHERE ct.clientid = :clientid AND it.stock_qty < ct.quantity
                `,
                {
                    replacements: {
                        clientid: client,
                    },
                    type: db.sequelize.QueryTypes.SELECT,
                },
            );
            const total = cart.reduce((acc, item) => {
                return acc + item.price * item.quantity;
            }, 0);
            return {
                statusCode: 200,
                message: 'Update cart successfully',
                data: {
                    cart,
                    total: parseFloat(total.toFixed(2)),
                    out_of_stock: itemOutOfStock,
                },
            };
        } catch (err) {
            throw new AppError(err.statusCode, err.message);
        }
    },
    deleteCart: async (client) => {
        try {
            await db.cart_item.destroy({
                where: {
                    clientid: client,
                },
            });
            return {
                statusCode: 200,
                message: 'Delete cart successfully',
            };
        } catch (err) {
            throw new AppError(err.statusCode, err.message);
        }
    },
    createCartItem: async (client, items) => {
        try {
            let error = [];
            for (const item of items) {
                const itemObj = await db.item.findOne({
                    where: {
                        uid: item.item,
                    },
                });
                if (item.quantity < 1)
                    error.push({
                        item: item.item,
                        message: 'Quantity must be greater than 0',
                    });
                if (!itemObj)
                    error.push({
                        item: item.item,
                        message: 'Item not found',
                    });
                const cartItem = await db.cart_item.findOne({
                    where: {
                        clientid: client,
                        itemid: item.item,
                    },
                });
                if (cartItem) {
                    if (
                        itemObj.stock_qty - cartItem.quantity - item.quantity <
                        0
                    )
                        error.push({
                            item: item.item,
                            message: 'Quantity is greater than stock quantity',
                        });
                    else {
                        cartItem.quantity += item.quantity;
                        await cartItem.save();
                    }
                } else {
                    if (itemObj.stock_qty - item.quantity < 0)
                        error.push({
                            item: item.item,
                            message: 'Quantity is greater than stock quantity',
                        });
                    else
                        await db.cart_item.create({
                            clientid: client,
                            itemid: item.item,
                            quantity: item.quantity,
                        });
                }
            }
            return {
                statusCode: 200,
                message: 'Create cart item successfully',
                data: error,
            };
        } catch (err) {
            throw new AppError(err.statusCode, err.message);
        }
    },
    getDiscount: async (client, code) => {
        try {
            const promotion = await db.promotion.findOne({
                where: {
                    code,
                },
            });
            if (!promotion) throw new AppError(404, 'Promotion not found');
            if (promotion.start_date > Date.now())
                throw new AppError(400, 'Promotion is not started yet');
            if (promotion.end_date < Date.now())
                throw new AppError(400, 'Promotion is ended');
            const cart = await db.sequelize.query(
                `
                SELECT it.name, quantity, link as primary_img, it.price, ct.itemid, categoryid
                FROM cart_items ct
                    JOIN items it ON ct.itemid = it.uid
                    LEFT JOIN item_imgs itm ON it.uid = itm.itemid AND itm.is_primary = 1
                WHERE ct.clientid = :clientid
                `,
                {
                    replacements: {
                        clientid: client,
                    },
                    type: db.sequelize.QueryTypes.SELECT,
                },
            );
            if (!cart.length) throw new AppError(400, 'Cart is empty');
            const itemOutOfStock = [];
            for (const item of cart) {
                const itemObj = await db.item.findOne({
                    where: {
                        uid: item.itemid,
                    },
                });
                if (itemObj.stock_qty - item.quantity < 0) {
                    cart.splice(cart.indexOf(item), 1);
                    itemOutOfStock.push(item);
                }
            }
            const condition = JSON.parse(promotion.condition);
            const discount_rate = promotion.discount_rate;
            const max_discount = promotion.max_discount;
            const type = promotion.type;
            let total = 0;
            let discount = 0;
            if (type === 'by total') {
                cart.forEach((item) => {
                    item.old_price = item.price;
                    item.new_price = item.price;
                    delete item.price;
                    total += item.new_price * item.quantity;
                    delete item.categoryid;
                });
                if (condition.total > total)
                    throw new AppError(400, 'Total is not enough get discount');
                discount = Math.min(total * discount_rate, max_discount);
                total -= discount;
            } else {
                let flag = false;
                cart.forEach((item) => {
                    item.old_price = item.price;
                    if (
                        (condition.item[0] === '*' &&
                            (condition.category[0] === '*' ||
                                condition.category.includes(
                                    item.categoryid,
                                ))) ||
                        condition.item.includes(item.itemid)
                    ) {
                        flag = true;
                        item.new_price = parseFloat(
                            (
                                item.price -
                                Math.min(
                                    item.price * discount_rate,
                                    max_discount,
                                )
                            ).toFixed(2),
                        );
                    } else {
                        item.new_price = item.price;
                    }
                    delete item.price;
                    discount +=
                        item.old_price * item.quantity -
                        item.new_price * item.quantity;
                    total += item.new_price * item.quantity;
                    delete item.categoryid;
                });
                if (!flag)
                    throw new AppError(
                        400,
                        'No item in cart is eligible for discount',
                    );
            }
            return {
                statusCode: 200,
                message: 'Get discount successfully',
                data: {
                    cart,
                    discount: parseFloat(discount.toFixed(2)),
                    total: parseFloat(total.toFixed(2)),
                    out_of_stock: itemOutOfStock,
                },
            };
        } catch (err) {
            throw new AppError(err.statusCode, err.message);
        }
    },
    checkout: async (
        clientid,
        { code, payment_method, receive_method, shipping_addressid },
    ) => {
        try {
            let cart;
            if (code) {
                cart = module.exports
                    .getDiscount(clientid, code)
                    .then((res) => res.data.cart)
                    .catch((err) => {
                        throw new AppError(err.statusCode, err.message);
                    });
            } else cart = (await module.exports.getCart(clientid)).data.cart;
            const address = db.client_address.findOne({
                where: {
                    uid: shipping_addressid,
                },
            });
            if (!address) throw new AppError(404, 'Address not found');
            if (address.clientid !== clientid)
                throw new AppError(400, 'Address is not belong to client');
            const order = await db.order.create({
                uid: v4(),
                clientid,
                payment_method,
                receive_method,
                shipping_addressid,
                order_date: Date.now(),
                promotion_code: code || null,
                order_status: 'Initiated',
            });
            cart.forEach(async (item) => {
                await db.order_item.create({
                    orderid: order.uid,
                    itemid: item.itemid,
                    quantity: item.quantity,
                    sales_price: item.old_price || item.price,
                });
                const itemObj = await db.item.findOne({
                    where: {
                        uid: item.itemid,
                    },
                });
                itemObj.stock_qty -= item.quantity;
                await itemObj.save();
            });
            await db.cart_item.destroy({
                where: {
                    clientid,
                },
            });
            return {
                statusCode: 200,
                message: 'Checkout successfully',
            };
        } catch (error) {
            throw new AppError(error.statusCode, error.message);
        }
    },
};

File: server_asp/src/api/cart/CartService.cs
using shieldtify.common;
using shieldtify.models;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using Microsoft.EntityFrameworkCore;
using Microsoft.AspNetCore.Mvc;
namespace shieldtify.api.cart
{
    public static class CartService
    {
        public static APIRes getCart(string clientID)
        {
            try
            {
                using var db = new ShieldtifyContext();
                var cart = db.CartItems.Where(i => i.Clientid.ToString() == clientID && i.Quantity <= i.Item.StockQty).Select(i => new
                {
                    name = i.Item.Name,
                    quantity = i.Quantity,
                    price = i.Item.Price,
                    primary_img = i.Item.ItemImgs.Where(im => im.IsPrimary).Select(im => im.Link).FirstOrDefault(),
                    itemid = i.Itemid
                }).ToList();
                var itemOutOfStock = db.CartItems.Where(i => i.Clientid.ToString() == clientID && i.Quantity > i.Item.StockQty).Select(i => new
                {
                    name = i.Item.Name,
                    quantity = i.Quantity,
                    price = i.Item.Price,
                    primary_img = i.Item.ItemImgs.Where(im => im.IsPrimary).Select(im => im.Link).FirstOrDefault(),
                    itemid = i.Itemid
                }).ToList();
                float total = 0;
                foreach (var item in cart)
                {
                    total += item.price * item.quantity;
                }
                return new APIRes(200, "Get cart successfully", new
                {
                    cart,
                    total = Math.Round(total, 2),
                    out_of_stock = itemOutOfStock
                });
            }
            catch (System.Exception)
            {
                throw;
            }
        }
        public static APIRes updateCart(string clientID, string itemID, int quantity)
        {
            try
            {
                using var db = new ShieldtifyContext();
                var cartItem = db.CartItems.Where(i => i.Clientid.ToString() == clientID && i.Itemid.ToString() == itemID).FirstOrDefault();
                if (cartItem == null)
                    return new APIRes(404, "Item not found in cart");
                if (quantity < 1)
                    db.CartItems.Remove(cartItem);
                else
                {
                    var item = db.Items.Where(i => i.Uid.ToString() == itemID).FirstOrDefault();
                    if (quantity > item.StockQty)
                        return new APIRes(400, "Quantity is greater than stock quantity");
                    else
                        cartItem.Quantity = quantity;
                }
                db.SaveChanges();
                var cart = db.CartItems.Where(i => i.Clientid.ToString() == clientID && i.Quantity <= i.Item.StockQty).Select(i => new
                {
                    name = i.Item.Name,
                    quantity = i.Quantity,
                    price = i.Item.Price,
                    primary_img = i.Item.ItemImgs.Where(im => im.IsPrimary).Select(im => im.Link).FirstOrDefault(),
                    itemid = i.Itemid
                }).ToList();
                var itemOutOfStock = db.CartItems.Where(i => i.Clientid.ToString() == clientID && i.Quantity > i.Item.StockQty).Select(i => new
                {
                    name = i.Item.Name,
                    quantity = i.Quantity,
                    price = i.Item.Price,
                    primary_img = i.Item.ItemImgs.Where(im => im.IsPrimary).Select(im => im.Link).FirstOrDefault(),
                    itemid = i.Itemid
                }).ToList();
                float total = 0;
                foreach (var item in cart)
                {
                    total += item.price * item.quantity;
                }
                return new APIRes(200, "Update cart successfully", new
                {
                    cart,
                    total = Math.Round(total, 2),
                    out_of_stock = itemOutOfStock
                });
            }
            catch (System.Exception)
            {

                throw;
            }
        }

        public static APIRes deleteCart(string clientID)
        {
            try
            {
                using var db = new ShieldtifyContext();
                var cartItems = db.CartItems.Where(i => i.Clientid.ToString() == clientID).ToList();
                if (cartItems == null)
                {
                    return new APIRes(404, "Cart not found");
                }
                db.CartItems.RemoveRange(cartItems);
                db.SaveChanges();
                return new APIRes(200, "Delete cart successfully");
            }
            catch (System.Exception)
            {
                throw;
            }
        }

        public static APIRes createCartItem(string clientID, List<Items> items)
        {
            try
            {
                using var db = new ShieldtifyContext();
                var error = new List<object>();
                foreach (var item in items)
                {
                    var itemObj = db.Items.Where(i => i.Uid.ToString() == item.item).FirstOrDefault();
                    if (item.quantity < 1)
                        error.Add(new { item.item, message = "Quantity must be greater than 0" });
                    if (itemObj == null)
                        error.Add(new { item.item, message = "Item not found" });
                    var cartItem = db.CartItems.Where(i => i.Clientid.ToString() == clientID && i.Itemid.ToString() == item.item).FirstOrDefault();
                    if (cartItem != null)
                    {
                        if (cartItem.Quantity + item.quantity > itemObj.StockQty)
                            error.Add(new { item.item, message = "Quantity is greater than stock quantity" });
                        else
                            cartItem.Quantity += item.quantity;
                    }
                    else
                    {
                        if (item.quantity > itemObj.StockQty)
                            error.Add(new { item.item, message = "Quantity is greater than stock quantity" });
                        else
                            db.CartItems.Add(new CartItem
                            {
                                Clientid = Guid.Parse(clientID),
                                Itemid = Guid.Parse(item.item),
                                Quantity = item.quantity
                            });
                    }
                }
                db.SaveChanges();
                return new APIRes(200, "Create cart item successfully", error);
            }
            catch (System.Exception)
            {
                throw;
            }
        }

        public static APIRes getDiscount(string clientID, [FromQuery] string code)
        {
            try
            {
                using var db = new ShieldtifyContext();
                var promotion = db.Promotions.Where(i => i.Code == code).FirstOrDefault();
                if (promotion == null)
                    return new APIRes(404, "Promotion not found");
                if (promotion.StartDate > DateTime.Now)
                    return new APIRes(400, "Promotion is not started yet");
                if (promotion.EndDate < DateTime.Now)
                    return new APIRes(400, "Promotion is ended");
                var cart = db.CartItems.Where(i => i.Clientid.ToString() == clientID && i.Quantity <= i.Item.StockQty).Select(i => new
                {
                    name = i.Item.Name,
                    quantity = i.Quantity,
                    price = i.Item.Price,
                    primary_img = i.Item.ItemImgs.Where(im => im.IsPrimary).Select(im => im.Link).FirstOrDefault(),
                    itemid = i.Itemid,
                    categoryid = i.Item.Categoryid
                }).ToList();
                var itemOutOfStock = db.CartItems.Where(i => i.Clientid.ToString() == clientID && i.Quantity > i.Item.StockQty).Select(i => new
                {
                    name = i.Item.Name,
                    quantity = i.Quantity,
                    price = i.Item.Price,
                    primary_img = i.Item.ItemImgs.Where(im => im.IsPrimary).Select(im => im.Link).FirstOrDefault(),
                    itemid = i.Itemid
                }).ToList();
                var condition = JsonConvert.DeserializeObject<JObject>(promotion.Condition);
                var discountRate = promotion.DiscountRate;
                var maxDiscount = promotion.MaxDiscount;
                var type = promotion.Type;
                float total = 0;
                float discount = 0;
                var items = new List<object>();
                if (type == "by total")
                {
                    foreach (var item in cart)
                    {
                        items.Add(new
                        {
                            item.itemid,
                            item.name,
                            item.quantity,
                            item.primary_img,
                            old_price = item.price,
                            new_price = item.price
                        });
                        total += item.price * item.quantity;
                    }
                    Console.WriteLine(total);
                    Console.WriteLine(int.Parse(condition["total"].ToString()));
                    if (int.Parse(condition["total"].ToString()) > total)
                        return new APIRes(400, "Total is not enough to get discount");
                    discount = (float)Math.Round(Math.Min(total * discountRate, maxDiscount), 2);
                    total -= discount;
                }
                else
                {
                    var flag = false;
                    foreach (var item in cart)
                    {
                        float newPrice = item.price;
                        if ((condition["item"][0].ToString() == "*" && (condition["category"][0].ToString() == "*" || condition["category"].ToObject<List<string>>().Contains(item.categoryid.ToString()))) || condition["item"].ToObject<List<string>>().Contains(item.itemid.ToString()))
                        {
                            items.Add(new
                            {
                                item.itemid,
                                item.name,
                                item.quantity,
                                item.primary_img,
                                old_price = item.price,
                                new_price = (float)Math.Round(item.price - Math.Min(item.price * discountRate, maxDiscount), 2)
                            });
                            flag = true;
                        }
                        else
                        {
                            items.Add(new
                            {
                                item.itemid,
                                item.name,
                                item.quantity,
                                item.primary_img,
                                old_price = item.price,
                                new_price = item.price
                            });
                        }
                        total += newPrice * item.quantity;
                        discount += Math.Min(newPrice * discountRate, maxDiscount);
                    }
                    if (!flag)
                        return new APIRes(400, "No item in cart is eligible for discount");
                }
                return new APIRes(200, "Get discount successfully", new
                {
                    cart = items,
                    discount = Math.Round(discount, 2),
                    total = Math.Round(total, 2),
                    out_of_stock = itemOutOfStock
                });
            }
            catch (System.Exception)
            {

                throw;
            }
        }
        public static APIRes checkout(string clientid, CheckoutBody checkoutBody)
        {
            try
            {
                var code = checkoutBody.code;
                var payment_method = checkoutBody.payment_method;
                var receive_method = checkoutBody.receive_method;
                var shipping_addressid = checkoutBody.shipping_addressid;
                using var db = new ShieldtifyContext();
                var cart = new List<object>();
                if (code != null)
                {
                    dynamic res = getDiscount(clientid, code);
                    if (res.statusCode != 200)
                        return res;
                    cart = (List<object>)res.data.cart;
                }
                else
                {
                    dynamic res = getCart(clientid);
                    if (res.statusCode != 200)
                        return res;
                    cart = (List<object>)res.data.cart;
                }
                var address = db.ClientAddresses.Where(i => i.Uid.ToString() == shipping_addressid).FirstOrDefault();
                if (address == null)
                    return new APIRes(404, "Address not found");
                if (address.Clientid.ToString() != clientid)
                    return new APIRes(400, "Address is not belong to client");
                var order = new Order
                {
                    Uid = Guid.NewGuid(),
                    Clientid = Guid.Parse(clientid),
                    PaymentMethod = payment_method,
                    ReceiveMethod = receive_method,
                    ShippingAddressid = Guid.Parse(shipping_addressid),
                    OrderDate = DateTime.Now,
                    PromotionCode = code,
                    OrderStatus = "Initiated"
                };
                db.Orders.Add(order);
                foreach (dynamic item in cart)
                {
                    string itemid = item.itemid.ToString();
                    var itemObj = db.Items.Where(i => i.Uid.ToString() == itemid).FirstOrDefault();
                    var orderItem = new OrderItem
                    {
                        Orderid = order.Uid,
                        Itemid = Guid.Parse(itemid),
                        Quantity = int.Parse(item.quantity.ToString()),
                        SalesPrice = item.old_price ?? item.price
                    };
                    db.OrderItems.Add(orderItem);
                    itemObj.StockQty -= int.Parse(item.quantity.ToString());
                }
                db.CartItems.RemoveRange(db.CartItems.Where(i => i.Clientid.ToString() == clientid));
                db.SaveChanges();
                return new APIRes(200, "Checkout successfully");
            }
            catch (System.Exception)
            {
                throw;
            }
        }
    }
}